### ISSUE TITLE
[MENFORCER-376] Add support for excludes/includes in requireJavaVendor rule

### ISSUE THREAD

**Issue by jira-importer** [apache/maven-enforcer#721]
**[Krosheninnikov Artem](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=kroartem)** opened **[MENFORCER-376](https://issues.apache.org/jira/browse/MENFORCER-376?redirect=false)** and commented

There was a suggestion here [1] to add includes/excludes support in requireJavaVendor rule. Right now it's not clear how it would work if you define the same vendor name in exclude and include lists but implementation can be more or less copied from BannedDependencies rule.

[1] https://issues.apache.org/jira/browse/MENFORCER-338?focusedCommentId=17169044&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-17169044

---

**Affects:** 3.0.0-M3

**Remote Links:**
- [GitHub Pull Request #86
](https://github.com/apache/maven-enforcer/pull/86)

**Comment by jira-importer** [comment]
**[Krosheninnikov Artem](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=kroartem)** commented

There are several questions:

1) It should not be possible to define one vendor both in includes and excludes, right?

2) If vendor name is not from includes but it's also not in excludes, what should happen?

[MattNelson](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=MattNelson), [Jeffrey Bennett](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jbennett2091), WDYT?

**Comment by jira-importer** [comment]
**[Jeffrey Bennett](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jbennett2091)** commented

The normal case for both an include and exclude being present would be
something like include-a-directory but exclude-file-in-directory.  In that
scenario, include and exclude make sense as they operate like an
all-but-this-one complement to each other.

That said, such a scenario does not exist here.  The vendors listed are
specific ones.

In short, should be either includes or excludes but not both.  If you have
either an include, or an exclude, or neither, it's
pretty straight forward.

On Sun, Jan 24, 2021, 9:45 AM Krosheninnikov Artem (Jira) <jira@apache.org>

**Comment by jira-importer** [comment]
**[Krosheninnikov Artem](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=kroartem)** commented

[Jeffrey Bennett](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jbennett2091), thanks for your input. I did current implementation as you described it in the first sentence. It indeed sounds counter-intuitive in a case with specific vendors. It's also good to get your reaction as PR was more or less reviewed could have reached master.

I'll rework the rule to support either excludes or includes, not both.

**Comment by jira-importer** [comment]
**[Matt Nelson](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=mattnelson)** commented

Is there a way to signal this issue should be a blocker for the next release 3.0.0 or 3.0.0-M4? The changes proposed here are non passive to the yet to be released changes from MENFORCER-338. I want to protect against a scenario where enforcer is released without this change, thus preventing it from being accepted due to being non passive.

**Comment by jira-importer** [comment]
**[Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte)** commented

Fixed in [3b47fa9627bdf96a9fb45aac31c655bc6cae5060](https://gitbox.apache.org/repos/asf?p=maven-enforcer.git;a=commit;h=3b47fa9627bdf96a9fb45aac31c655bc6cae5060)
